spec/formatting.md

```diff
@@ -57,6 +57,23 @@ nor be made available to _function handlers_.
 > _declarations_ affecting _variables_ referenced by that _expression_
 > have already been evaluated in the order in which the relevant _declarations_
 > appear in the _message_.
+>
+> Implementations and users creating custom functions SHOULD avoid
+> creating _function handlers_ that mutate external program state.
+>
+
+> [!NOTE]
+>
+> Function handlers can also be written that depend on external mutable state
+> (for example, the current date and time).
+> Lazy evaluation might involve evaluating the same _expression_ multiple times
+> (call-by-name) or evaluating every expression at most once (call-by-need).
```

Comment:
I really think it's a mistake to allow the same expression to ever evaluate to a different value. This is something that [CEL](https://github.com/google/cel-spec/blob/master/doc/langdef.md#evaluation) (to pick a roughly analogous technology) gets _very_ right.